keylime/user_utils.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/python3
+
+'''
+SPDX-License-Identifier: Apache-2.0
+Copyright 2017 Massachusetts Institute of Technology.
+'''
+
+import os
+import grp
+import pwd
+
+from keylime import keylime_logging
+
+# Configure logger
+logger = keylime_logging.init_logging('privileges')
+
+
+def string_to_uidgid(userandgroup):
+    """ Translate the userandgroup string to uid and gid.
+        The userandgroup parameter must be a string of the format '<user>[:<group>]'.
+        User and group can be strings or integers. If no group is given, -1 will be
+        returned for gid.
+    """
+    params = userandgroup.split(':')
+    if len(params) > 2:
+        return None, None, "User and group '%s' are in wrong format. Expected <user>[:group]" % userandgroup
+
+    gid = -1
+    if len(params) == 2:
+        if params[1].isnumeric():
+            gid = int(params[1])
+        else:
+            try:
+                gr = grp.getgrnam(params[1])
+                gid = gr.gr_gid
+            except Exception as e:
+                return None, None, 'Could not resolve group %s: %s' % (params[1], str(e))
+
+    if params[0].isnumeric():
+        uid = int(params[0])
+    else:
+        try:
+            passwd = pwd.getpwnam(params[0])
+            uid = passwd.pw_uid
+        except Exception as e:
+            return None, None, 'Could not resolve user %s: %s' % (params[0], str(e))
+
+    return uid, gid, None
+
+
+def change_uidgid(userandgroup):
+    """ Change uid and gid of the current process.
+        The userandgroup parameter must be a string of the format
+        '<user>[:<group>]'. User and group can be strings or integers.
+    """
+    uid, gid, err = string_to_uidgid(userandgroup)
+    if err:
+        return err
+
+    # First change group and then user
+    if gid >= 0:
```

Comment:
I can see how `gid == -1` is to maintain the same user, but for me is a more natural way that `if uid is not None` and `if gid is not None` imply to maintain the same `uid` or `gid` independently.  This open the door to have `uid`, `uid:`, `:gid`, `uid:gid` combinations in the `cloud_agent.run_as` parameter.

All this can be covered updating the parser, and here treated as a two independent choices.

Also a test for those situation will make easy the refactoring of the code.

Reply:
Added unittest and swtiched to use `None`.